### DISCUSSION TITLE
How do I create a unittest for a servicelayer which uses the client?

### DISCUSSION THREAD

**Discussion by MartijnStraatman** [fabric8io/kubernetes-client#5050] [Q&A] [answered]
How do I mock the client in a proper way so I can unittest my dependent services?

For example

```
    private MixedOperation<io.clastix.capsule.v1beta1.Tenant,
    KubernetesResourceList<io.clastix.capsule.v1beta1.Tenant>, Resource<io.clastix.capsule.v1beta1.Tenant>> tenantClient() {
        return  kubernetesClient.resources(io.clastix.capsule.v1beta1.Tenant.class);
    }
 ```
 
 If I call `tenantClient` I want to mock a response. For example a Tenant Resource or a list for resources?

I tried several things with Mockito. Maybe someone can help me a bit?

**Comment by rohanKanojia** [comment]
@MartijnStraatman : Have you tried [Fabric8 Kubernetes Mock Server](https://search.maven.org/artifact/io.fabric8/kubernetes-server-mock/6.5.1/jar)? You can see an example of mocking custom resource call [here](https://github.com/fabric8io/kubernetes-client/blob/master/kubernetes-tests/src/test/java/io/fabric8/kubernetes/client/mock/TypedCustomResourceApiTest.java#L60).

There is also an option of setting up test etcd/kubernetes apiserver environment using [jenvtest](https://github.com/java-operator-sdk/jenvtest) . See [blogpost](https://csviri.medium.com/introducing-jenvtest-kubernetes-api-server-tests-made-easy-for-java-4d02a9bb26d4)

**Comment by MartijnStraatman** [comment]
I did. The CRD's I am using is not something I have created myself. I am just wondering how I can test my own service logic against the kubernetes responses. So how can I intregrate the Fabric8 Kubernetes Mock Server in my unittest as the service layer will use that Mock Server?

**Reply by rohanKanojia** [reply, marked as answer]
~You can read my blogpost to get some idea about this [How to write tests with Fabric8 Kubernetes Client](https://developers.redhat.com/articles/2023/01/24/how-write-tests-fabric8-kubernetes-client)~

**Comment by MartijnStraatman** [comment]
Thx, worked!

**Reply by rohanKanojia** [reply]
Sorry, I didn't understand your question properly. What do you mean by this?
> how can I intregrate the Fabric8 Kubernetes Mock Server in my unittest as the service layer will use that Mock Server
